src/chain/chain.rs

```diff
@@ -870,17 +871,44 @@ impl Chain {
 
     /// Returns `true` if the `proof_chain` contains a key we have in `their_keys` and that key is
     /// for a prefix compatible with proof_chain prefix.
-    pub fn check_trust(&self, proof_chain: &SectionProofChain) -> bool {
+    pub fn check_trust(&self, proof_chain: &SectionProofChain) -> Result<(), ProofError> {
         let last_prefix = proof_chain.last_public_key_info().prefix();
-        let filtered_keys: BTreeSet<_> = self
+        let known_key_infos: BTreeSet<_> = self
             .state
             .get_their_keys_info()
             .filter(|&(pfx, _)| last_prefix.is_compatible(pfx))
             .map(|(_, info)| info)
             .collect();
-        proof_chain
+
+        if proof_chain
+            .all_key_infos()
+            .any(|proof_key_info| known_key_infos.contains(proof_key_info))
+        {
+            return Ok(());
+        }
+
+        let (min_known_version, max_known_version) = known_key_infos
+            .iter()
+            .map(|known_key_info| known_key_info.version())
+            .minmax()
+            .into_option()
+            .ok_or(ProofError::Invalid)?;
+        let (min_proof_version, max_proof_version) = proof_chain
             .all_key_infos()
-            .any(|key_info| filtered_keys.contains(key_info))
+            .map(|proof_key_info| proof_key_info.version())
+            .minmax()
+            .into_option()
+            .expect("empty proof");
```

Comment:
Would it make sense to avoid the unecessary iterator and expect and use something like:
```
let (min_proof_version, max_proof_version) = (proof_chain.genesis_public_key_info().version(), proof_chain.last_public_key_info().version()); // the genesis function is missing but probably should be added.
```

Reply:
Could strengthen that expectation I guess. It seem to be a sensible expectation.

Reply:
I think we can actually do better. We can modify `SectionProofChain` so that expectation is statically guaranteed. Probably not in this PR though.

Reply:
Would it make sense to still use that expectation here though?